partiql-spi/src/main/kotlin/org/partiql/spi/utils/IntervalUtils.kt

```diff
@@ -0,0 +1,159 @@
+    private const val NANOS_PER_SECOND = 1_000_000_000L
+
+    /**
+     * Divides an interval by a numeric value.
+     *
+     * @param interval The interval type to divide
+     * @return A function that takes an interval Datum and BigDecimal divisor, returns the divided interval
+     */
+    fun intervalDivide(interval: PType): ((Datum, Number) -> Datum) {
+        return when (interval.intervalCode) {
+            IntervalCode.YEAR,
+            IntervalCode.MONTH,
+            IntervalCode.YEAR_MONTH -> { i, number ->
+                // get total months from the interval as Long Type as it can hold the interval with max precision = 9.
+                val totalMonths: Long = i.years * MONTHS_PER_YEAR + i.months
+                val totalMonthsInDouble = totalMonths / number.toDouble()
+                val years = (totalMonthsInDouble.toLong() / MONTHS_PER_YEAR).toInt()
+                val months = (totalMonthsInDouble.toLong() % MONTHS_PER_YEAR).toInt()
+                Datum.intervalYearMonth(years, months, interval.precision)
+            }
+
+            IntervalCode.DAY,
+            IntervalCode.DAY_HOUR,
+            IntervalCode.DAY_MINUTE,
+            IntervalCode.DAY_SECOND,
+            IntervalCode.HOUR,
+            IntervalCode.HOUR_MINUTE,
+            IntervalCode.HOUR_SECOND,
+            IntervalCode.MINUTE,
+            IntervalCode.MINUTE_SECOND,
+            IntervalCode.SECOND -> { i, number ->
+                // When float/double is converted to BigDecimal, NumberUtils.bigDecimalOf will introduce inaccuracy for double due to the nature of approximate type
+                // For example, 0.1 could be 0.100000000000000111 which leads to computing inaccuracy.
+                val numberInBigDecimal = NumberUtils.bigDecimalOf(number)
+                val totalSecondsInBigDecimal = toSeconds(i)
+                val resultInBigDecimal = totalSecondsInBigDecimal.divide(numberInBigDecimal, NANO_MAX_PRECISION, INTERVAL_ROUNDING_MODE)
+                fromSecond(resultInBigDecimal, interval.precision, getFractionPrecision(interval))
+            }
+
+            else -> throw IllegalArgumentException("Unable to calculate division for INTERVAL expression")
+        }
+    }
+
+    /**
+     * Multiplies an interval by a numeric value.
+     *
+     * @param interval The interval type to multiply
+     * @return A function that takes an interval Datum and BigDecimal multiplier, returns the multiplied interval
+     */
+    fun intervalMultiply(interval: PType): ((Datum, Number) -> Datum) {
+        return when (interval.intervalCode) {
+            IntervalCode.YEAR,
+            IntervalCode.MONTH,
+            IntervalCode.YEAR_MONTH -> { i, number ->
+                // get total months from the interval as Long Type as it can hold the interval with max precision = 9.
+                val totalMonths: Long = i.years * MONTHS_PER_YEAR + i.months
+                val totalMonthsInDouble = totalMonths * number.toDouble()
```

Comment:
Similarly here. We should not convert exact number representations to approximate if `number` is already exact (int, decimal, etc).